File: NeetCodeIO/347TopKFrequentElements/TopKFrequentElements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Bucket Sort -> Time:O(n), Space:O(n)
        count = {}
        freq = [[] for i in range(len(nums)+1)]
        ans = []
        for i in nums:
            if i not in count:
                count[i] = 0
            count[i] += 1
        for n,c in count.items():
            freq[c].append(n)
        for i in range(len(nums), 0, -1):
            for n in freq[i]:
                ans.append(n)
                if len(ans) == k:
                    return ans
        return ans

File: NeetCodeIO/347TopKFrequentElements/TopKFrequentElements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Bucket Sort -> Time:O(n), Space:O(n)
        map<int,int> m;
        vector<vector<int>> freq;
        vector<int> ans;
        int n = nums.size();
        for (int i = 0; i <= n; i++) {
            vector<int> temp;
            freq.push_back(temp);
        }
        for (int i = 0; i < n; i++) {
            m[nums[i]]++;
        }
        for (auto i:m) {
            freq[i.second].push_back(i.first);
        }
        for (int i = n; i > 0; i--) {
            for (int j = 0; j < freq[i].size(); j++) {
                ans.push_back(freq[i][j]);
                if (ans.size() == k) {
                    return ans;
                }
            }
        }
        return ans;
    }
};
